pkg/csi/cinder/controllerserver.go

```diff
@@ -315,52 +346,130 @@ func (cs *controllerServer) ControllerUnpublishVolume(ctx context.Context, req *
 	return &csi.ControllerUnpublishVolumeResponse{}, nil
 }
 
+type CloudsStartingToken struct {
+	CloudName string `json:"cloud"`
+	Token     string `json:"token"`
+}
+
 func (cs *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
 	klog.V(4).Infof("ListVolumes: called with %+#v request", req)
 
 	if req.MaxEntries < 0 {
 		return nil, status.Errorf(codes.InvalidArgument, "[ListVolumes] Invalid max entries request %v, must not be negative ", req.MaxEntries)
 	}
+
 	maxEntries := int(req.MaxEntries)
 
-	vlist, nextPageToken, err := cs.Cloud.ListVolumes(maxEntries, req.StartingToken)
-	if err != nil {
-		klog.Errorf("Failed to ListVolumes: %v", err)
-		if cpoerrors.IsInvalidError(err) {
-			return nil, status.Errorf(codes.Aborted, "[ListVolumes] Invalid request: %v", err)
-		}
-		return nil, status.Errorf(codes.Internal, "ListVolumes failed with error %v", err)
+	var err error
+	var cloudsToken = CloudsStartingToken{
+		CloudName: "",
+		Token:     "",
 	}
 
-	ventries := make([]*csi.ListVolumesResponse_Entry, 0, len(vlist))
-	for _, v := range vlist {
-		ventry := csi.ListVolumesResponse_Entry{
-			Volume: &csi.Volume{
-				VolumeId:      v.ID,
-				CapacityBytes: int64(v.Size * 1024 * 1024 * 1024),
-			},
+	cloudsNames := maps.Keys(cs.Clouds)
+	sort.Strings(cloudsNames)
+
+	currentCloudName := cloudsNames[0]
+	if req.StartingToken != "" {
+		err = json.Unmarshal([]byte(req.StartingToken), &cloudsToken)
+		if err != nil {
+			return nil, status.Errorf(codes.Aborted, "[ListVolumes] Invalid request: Token invalid")
 		}
+		currentCloudName = cloudsToken.CloudName
+	}
+
+	startingToken := cloudsToken.Token
+	var cloudsVentries []*csi.ListVolumesResponse_Entry
+	var vlist []volumes.Volume
+	var nextPageToken string
 
-		status := &csi.ListVolumesResponse_VolumeStatus{}
-		status.PublishedNodeIds = make([]string, 0, len(v.Attachments))
-		for _, attachment := range v.Attachments {
-			status.PublishedNodeIds = append(status.PublishedNodeIds, attachment.ServerID)
+	startIdx := 0
+	for idx := 0; idx < len(cloudsNames); idx++ {
```

Comment:
Why not iterate using `range()`?

Reply:
0128e96cebabcd2c44efce3307802c834026392b